x/ibc/03-connection/handshake.go

```diff
@@ -0,0 +1,416 @@
+		NextTimeout:        man.man.protocol.Value([]byte(parent.id + "/timeout")).Integer(state.Dec),
+
+		// CONTRACT: counterparty must be filled by the caller
+	}
+}
+
+func (man CounterpartyHandshaker) Object(id string) CounterHandshakeObject {
+	return CounterHandshakeObject{
+		CounterObject: man.man.Object(id),
+
+		State:              man.man.protocol.Value([]byte(id + "/state")).Enum(),
+		CounterpartyClient: man.man.protocol.Value([]byte(id + "/counterpartyClient")).String(),
+		NextTimeout:        man.man.protocol.Value([]byte(id + "/timeout")).Integer(state.Dec),
+	}
+}
+
+func (man Handshaker) create(ctx sdk.Context, id string, connection Connection, counterpartyClient string) (obj HandshakeObject, err error) {
+	cobj, err := man.man.create(ctx, id, connection, HandshakeKind)
+	if err != nil {
+		return
+	}
+	obj = man.Object(cobj)
+	obj.CounterpartyClient.Set(ctx, counterpartyClient)
+	obj.Counterparty = man.counterparty.Object(connection.Counterparty)
+	return obj, nil
+}
+
+func (man Handshaker) query(ctx sdk.Context, id string) (obj HandshakeObject, err error) {
+	cobj, err := man.man.query(ctx, id, HandshakeKind)
+	if err != nil {
+		return
+	}
+	obj = man.Object(cobj)
+	obj.Counterparty = man.counterparty.Object(obj.GetConnection(ctx).Counterparty)
+	return
+}
+
+func (obj HandshakeObject) remove(ctx sdk.Context) {
+	obj.Object.remove(ctx)
+	obj.State.Delete(ctx)
+	obj.CounterpartyClient.Delete(ctx)
+	obj.NextTimeout.Delete(ctx)
+}
+
+func assertTimeout(ctx sdk.Context, timeoutHeight uint64) error {
+	if uint64(ctx.BlockHeight()) > timeoutHeight {
+		return errors.New("timeout")
+	}
+
+	return nil
+}
+
+// Using proofs: none
+func (man Handshaker) OpenInit(ctx sdk.Context,
+	id string, connection Connection, counterpartyClient string, nextTimeoutHeight uint64,
+) (HandshakeObject, error) {
+	// man.Create() will ensure
+	// assert(get("connections/{identifier}") === null) and
+	// set("connections{identifier}", connection)
+	obj, err := man.create(ctx, id, connection, counterpartyClient)
+	if err != nil {
+		return HandshakeObject{}, err
+	}
+
+	obj.NextTimeout.Set(ctx, nextTimeoutHeight)
+	obj.State.Set(ctx, Init)
+
+	return obj, nil
+}
+
+// Using proofs: counterparty.{connection,state,nextTimeout,counterpartyClient, client}
+func (man Handshaker) OpenTry(ctx sdk.Context,
+	proofs []commitment.Proof,
+	id string, connection Connection, counterpartyClient string, timeoutHeight, nextTimeoutHeight uint64,
+) (obj HandshakeObject, err error) {
+	obj, err = man.create(ctx, id, connection, counterpartyClient)
+	if err != nil {
+		return
+	}
+
+	ctx, err = obj.Context(ctx, connection.Path, proofs)
+	if err != nil {
+		return
+	}
+
+	err = assertTimeout(ctx, timeoutHeight)
+	if err != nil {
+		return
+	}
+
+	if !obj.Counterparty.State.Is(ctx, Init) {
+		err = errors.New("counterparty state not init")
+		return
+	}
+
+	if !obj.Counterparty.Connection.Is(ctx, Connection{
```

Comment:
I think this abstraction is OK but `Is` is not a very indicative name. This involves checking a proof presumably?

Reply:
Have to rename it to `Verify`, sorry